enhancements/dns/externaldns-operator.md

```diff
@@ -0,0 +1,477 @@
+
+The following tasks would be completed as a part of this enhancement:
+
+* Create a GitHub repository to host the operator source code in OpenShift's Github org,
+with the hopes of moving the operator's source code into an upstream k8s-sigs repository in the future.
+* Leverage popular frameworks and libraries , i.e. controller-runtime and kubebuilder, to simplify development
+of the operator and align with upstream communities.
+* Manage dependencies through [Go Modules](https://blog.golang.org/using-go-modules).
+* Create user and developer documentation for running ExternalDNS operator (on both OpenShift and vanilla Kubernetes).
+* Create tooling to simplify building, running, testing, etc. the operator.
+* Create tests that reduce bugs in the code and allow for continuous integration (tests should be compatible
+with OpenShift CI, in addition to vanilla kubernetes environments running in public clouds).
+* Integrate the operator with the OpenShift toolchain, i.e. [openshift/release](https://github.com/openshift/release),
+which would be required for productization.
+* Provide manifests for quickly deploying the operator (including a minimal ExternalDNS Operator CR manifest).
+* Create tooling to simplify ExternalDNS troubleshooting. For example, add support for the Operator to
+OpenShift [must-gather](https://github.com/openshift/must-gather).
+* Integrate ExternalDNS with the OpenShift monitoring toolchain. For example, add
+[metrics support](https://github.com/openshift/cluster-monitoring-operator) for the operator and its operand(s).
+
+
+In addition, the following functionality is expected as part of the operator (note that this is not an exhaustive list):
+
+* Introduce a CRD to be used by the operator to manage ExternalDNS and its dependencies.
+* Common fields in spec that define the configuration of ExternalDNS and any dependent resources,
+i.e. a secret resource containing credentials for the necessary cloud provider.
+* Common fields in spec that define how to deploy ExternalDNS.
+* Common fields in status that expose the current health & version information of ExternalDNS operands.
+* The operator will be capable of observing other resources to perform basic sequencing required for OpenShift
+integration.
+* The operator will be capable of deploying multiple instances of ExternalDNS (operands), which is necessary when
+a common domain is shared across DNS zones, or when using multiple DNS Providers. See ExternalDNS' upstream
+[AWS public/private zone tutorial docs](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/public-private-route53.md)
+as an example.
+
+Here is what a CRD instance for ExternalDNS Operator may look like:
+
+```yaml
+apiVersion: operator.externaldns.k8s.io/v1alpha1
+kind: externaldns
+metadata:
+  name: default
+  namespace: my-externaldns-namespace
+spec:
+  DNSZones:
+    - my-public-zone
+    - my-private-zone
```

Comment:
In the case of OpenShift, the cluster config identifies zones alternatively by id or by tags.  What is being specified in the ExternalDNS CR?  Are these domains, platform-specific identifiers, or what?  How do we map these to zones defined in the cluster config?  Will the ExternalDNS operator read the cluster config, look up the hosted zones, and map these zones from the ExternalDNS CR to the provider's hosted zones?  Can the example ExternalDNS deployment below be updated to show how the zones specified in the ExternalDNS CR map to the deployment?

Reply:
I elaborated on this in the enhancement. Thanks!

Reply:
I'm still unsure as to how this is supposed to be configured.  You're referring to the following, right?

> The `zones` field contains a list of DNS zone IDs. On OpenShift, the ExternalDNS Operator could read
> ￼the cluster DNS config resource and extract the available public/private zone IDs. Note that a zone lookup may need
> ￼to be performed by the ExternalDNS Operator in order to determine the ID of any zones that are specified
> ￼via a resource tag in the cluster DNS config resource. The ExternalDNS operator will then take the provided zone IDs
> ￼and map them to the given provider's complete hosted zone ID string.
￼
With OpenShift, the cluster config can identify a DNS zone by id or by tags.  Suppose a user (or, in the future, the installer itself) has the cluster DNS config at hand and wants to create an ExternalDNS CR.  Then, does the user (or installer) need to look up the id for every zone that were specified in the cluster config using tags?  Does the user leave `zones` empty and let the operator fill it in from the cluster DNS config?  How does this work?

Reply:
I've elaborated on this section again...I hope it's clear now! I'll also answer inline just so we are on the same page 

> Does the user leave zones empty and let the operator fill it in from the cluster DNS config? 
yes

> Suppose a user (or, in the future, the installer itself) has the cluster DNS config at hand and wants to create an ExternalDNS CR. Then, does the user (or installer) need to look up the id for every zone that were specified in the cluster config using tags?

Day 2: User
Day 1: Installer